torch/_functorch/partitioners.py

```diff
@@ -833,22 +842,39 @@ def get_node_weight(node) -> int:
             return mem_sz * 2
 
     nx_graph = nx.DiGraph()
+    edge_set = set()
+
+    # NOTE: `nx_graph.add_edge` inherently doesn't check whether the edge already exists
+    # (it will just overwrite any existing). We use our own set to prevent duplicate edges.
+    def add_edge(u, v, capacity):
+        assert (u, v) not in edge_set
+        nx_graph.add_edge(u, v, capacity=capacity)
+        edge_set.add((u, v))
+
     for node in full_bw_graph.nodes:
         if node.op == 'output':
             continue
 
         if node in required_bw_nodes:
-            nx_graph.add_edge(node.name + "_in", "sink", capacity=math.inf)
+            add_edge(node.name + "_in", "sink", capacity=math.inf)
             continue
 
         if _is_primal(node) or _is_fwd_seed_offset(node):
-            nx_graph.add_edge("source", node.name + "_in", capacity=math.inf)
-
-        # If a node can't be recomputed (too expensive or involves randomness),
-        # we prevent it from being recomputed by adding an inf edge to the source
-        # We only need to ban nodes in the fw pass, as those are the only ones that would be recomputed.
-        if ban_recomputation(node) and node in required_fw_nodes:
-            nx_graph.add_edge("source", node.name + "_in", capacity=math.inf)
+            add_edge("source", node.name + "_in", capacity=math.inf)
+
+        if node in required_fw_nodes:
+            if must_recompute(node):
+                # If user explicitly say they want to recompute a node, we honor it
+                # by adding an 0-capacity edge from the source and an inf edge to the sink.
+                # This way, X_in node is guaranteed to be part of the subgraph that contains "sink"
+                # after the cut, thus guaranteeing that X op will be recomputed.
+                add_edge("source", node.name + "_in", capacity=0)
+                add_edge(node.name + "_in", "sink", capacity=math.inf)
```

Comment:
mic-cut partitioner is supposed to give better cut on top of user-provided annotation. So, we should maybe not do this.

But if you want to do this deliberately (because FSDP does something funky), a better way might be to just have another config flag that respects the user AC annotation. In that case, you can just check `if flag_enabled and must_recompile()` without adding `prefer_recompute` or other logic.